File: mbox/android/app/src/main/assets/spider.js
// TVBox JS Spider 测试脚本
// 用于验证爬虫系统是否正常工作

var Spider = {
    init: function (extend) {
        console.log('JS Spider init: ' + extend);
        return true;
    },
    
    home: function (filter) {
        console.log('JS Spider home: filter=' + filter);
        var result = {
            'class': [
                {'type_id': '1', 'type_name': '电影'},
                {'type_id': '2', 'type_name': '电视剧'},
                {'type_id': '3', 'type_name': '综艺'},
                {'type_id': '4', 'type_name': '动漫'}
            ],
            'filters': {}
        };
        return JSON.stringify(result);
    },
    
    category: function (params) {
        console.log('JS Spider category: ' + JSON.stringify(params));
        var tid = params.tid;
        var pg = params.page || 1;
        var filter = params.filter || false;
        var result = {
            'list': [
                {
                    'vod_id': 'test_' + tid + '_' + pg,
                    'vod_name': '测试视频_' + tid + '_' + pg,
                    'type_name': '测试分类',
                    'vod_pic': 'https://via.placeholder.com/300x450',
                    'vod_remarks': 'HD',
                    'vod_year': '2024',
                    'vod_area': '中国',
                    'vod_director': '测试导演',
                    'vod_actor': '测试演员',
                    'vod_content': '这是一个测试视频的简介',
                    'vod_play_from': 'test',
                    'vod_play_url': '第 1 集$http://example.com/video1.mp4#第 2 集$http://example.com/video2.mp4'
                }
            ],
            'page': parseInt(pg),
            'pagecount': 10,
            'limit': 20,
            'total': 200
        };
        return JSON.stringify(result);
    },
    
    detail: function (params) {
        console.log('JS Spider detail: ' + JSON.stringify(params));
        var id = params.id;
        var result = {
            'list': [
                {
                    'vod_id': id,
                    'vod_name': '测试视频详情_' + id,
                    'type_name': '测试分类',
                    'vod_pic': 'https://via.placeholder.com/300x450',
                    'vod_remarks': 'HD',
                    'vod_year': '2024',
                    'vod_area': '中国',
                    'vod_director': '测试导演',
                    'vod_actor': '测试演员',
                    'vod_content': '这是一个测试视频的简介',
                    'vod_play_from': 'test',
                    'vod_play_url': '第 1 集$http://example.com/video1.mp4#第 2 集$http://example.com/video2.mp4#第 3 集$http://example.com/video3.mp4'
                }
            ]
        };
        return JSON.stringify(result);
    },
    
    play: function (params) {
        console.log('JS Spider play: ' + JSON.stringify(params));
        var flag = params.flag;
        var id = params.id;
        var vipFlags = params.vipFlags || '';
        
        // 返回播放地址
        var result = {
            'parse': 0,
            'url': 'http://example.com/video.mp4',
            'header': {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
        };
        return JSON.stringify(result);
    },
    
    search: function (params) {
        console.log('JS Spider search: ' + JSON.stringify(params));
        var wd = params.wd;
        var quick = params.quick || false;
        var result = {
            'list': [
                {
                    'vod_id': 'search_' + wd,
                    'vod_name': '搜索结果：' + wd,
                    'type_name': '测试分类',
                    'vod_pic': 'https://via.placeholder.com/300x450',
                    'vod_remarks': 'HD'
                }
            ]
        };
        return JSON.stringify(result);
    },
    
    destroy: function () {
        console.log('JS Spider destroyed');
        return true;
    }
};

// 全局函数包装
function home(filter) { return Spider.home(filter); }
function category(params) { return Spider.category(params); }
function detail(params) { return Spider.detail(params); }
function play(params) { return Spider.play(params); }
function search(params) { return Spider.search(params); }
function init(extend) { return Spider.init(extend); }
function destroy() { return Spider.destroy(); }

File: mbox/lib/crawler/spider.dart
import 'dart:convert';
import '../models/vod.dart';
import '../models/vod_config.dart';
import '../network/okhttp_client.dart';
import '../utils/log_utils.dart';

/// Spider 抽象基类
/// 参考 SPIDER.md 中的规格实现
abstract class Spider {
  String siteKey = '';

  /// 初始化
  /// 在 Spider 实例创建后调用一次
  Future<void> init(dynamic context, String extend) async {
    Log.d('Spider init: extend=$extend');
  }

  /// 首页分类
  /// 用户进入首页时调用
  Future<Map<String, dynamic>> homeContent(bool filter) async {
    return {'class': [], 'filters': {}};
  }

  /// 首页推荐影片
  /// 首页分类加载完成后调用
  Future<Map<String, dynamic>> homeVideoContent() async {
    return {'list': []};
  }

  /// 分类列表
  /// 用户点击分类或切换筛选条件时调用
  Future<Map<String, dynamic>> categoryContent(
    String tid,
    String pg,
    bool filter,
    Map<String, String> extend,
  ) async {
    return {'list': [], 'pagecount': 1};
  }

  /// 影片详情
  /// 用户点击影片卡片时调用
  Future<Map<String, dynamic>> detailContent(List<String> ids) async {
    return {'list': []};
  }

  /// 搜索
  /// 用户输入关键字搜索时调用
  Future<Map<String, dynamic>> searchContent(
    String key,
    bool quick, {
    String? pg,
  }) async {
    return {'list': []};
  }

  /// 播放解析
  /// 用户选择集数准备播放时调用
  Future<Map<String, dynamic>> playerContent(
    String flag,
    String id,
    List<String> vipFlags,
  ) async {
    return {'url': '', 'parse': 0};
  }

  /// 直播频道列表
  /// 加载直播源时调用
  Future<String> liveContent(String url) async {
    return '';
  }

  /// 本地代理
  /// 本地 HTTP 代理服务器收到请求时调用
  Future<List<dynamic>> proxy(Map<String, String> params) async {
    return [];
  }

  /// 自定义动作
  /// UI 层调用特定自定义指令时调用
  Future<Map<String, dynamic>> action(String actionStr) async {
    return {};
  }

  /// 手动检查视频格式
  /// 返回 true 时，框架会在 WebView 中拦截 URL 后调用 isVideoFormat
  Future<bool> manualVideoCheck() async {
    return false;
  }

  /// 检查是否为视频格式
  /// 判断指定 URL 是否为有效的直接媒体 URL
  Future<bool> isVideoFormat(String url) async {
    // 常见的视频格式
    final videoPatterns = [
      r'\.m3u8',
      r'\.mp4',
      r'\.flv',
      r'\.avi',
      r'\.mkv',
      r'\.wmv',
      r'\.mov',
      r'\.webm',
      r'\.mpd',
    ];
    
    for (var pattern in videoPatterns) {
      if (RegExp(pattern, caseSensitive: false).hasMatch(url)) {
        return true;
      }
    }
    
    return false;
  }

  /// 销毁
  /// 配置重新加载或应用清理缓存时调用
  Future<void> destroy() async {
    Log.d('Spider destroy');
  }

  /// 辅助方法：将结果转换为 JSON 字符串
  String toJson(Map<String, dynamic> data) {
    return json.encode(data);
  }

  /// 辅助方法：从 JSON 字符串解析
  Map<String, dynamic> fromJson(String jsonStr) {
    return json.decode(jsonStr);
  }

  /// 辅助方法：创建 Vod 对象
  Map<String, dynamic> createVod({
    required String vodId,
    required String vodName,
    String? vodPic,
    String? vodRemarks,
    String? typeName,
    String? vodYear,
    String? vodArea,
    String? vodDirector,
    String? vodActor,
    String? vodContent,
    String? vodPlayFrom,
    String? vodPlayUrl,
    String? vodTag,
  }) {
    return {
      'vod_id': vodId,
      'vod_name': vodName,
      'vod_pic': vodPic,
      'vod_remarks': vodRemarks,
      'type_name': typeName,
      'vod_year': vodYear,
      'vod_area': vodArea,
      'vod_director': vodDirector,
      'vod_actor': vodActor,
      'vod_content': vodContent,
      'vod_play_from': vodPlayFrom,
      'vod_play_url': vodPlayUrl,
      'vod_tag': vodTag,
    };
  }

  /// 辅助方法：创建 Class 对象
  Map<String, dynamic> createClass({
    required String typeId,
    required String typeName,
    String? typeFlag,
  }) {
    return {
      'type_id': typeId,
      'type_name': typeName,
      'type_flag': typeFlag,
    };
  }

  /// 辅助方法：创建 Filter 对象
  Map<String, dynamic> createFilter({
    required String key,
    required String name,
    required List<Map<String, dynamic>> value,
    String? init,
  }) {
    return {
      'key': key,
      'name': name,
      'value': value,
      'init': init,
    };
  }

  /// 辅助方法：繁简转换（如果需要）
  Future<String> traditionalToSimplified(String text) async {
    // TODO: 集成繁简转换库
    return text;
  }

  /// 辅助方法：网络请求
  Future<String> fetch(String url, {Map<String, String>? headers}) async {
    return await OkHttpUtils.get(url, headers: headers);
  }
}
